webpack/assets/javascripts/react_app/routes/RoutingService.js

```diff
@@ -0,0 +1,42 @@
+let currentLocation = null;
+
+/**
+ * a Helper function for rendering a route
+ * @param {Function} renderFn - a component's rendering function
+ * @param {Object} props - routing props
+ * @param {Function} beforeRenderingCallback - a callback to run before the the rendering, if it returns false the rendering will terminate
+ */
+export const renderRoute = (
```

Comment:
this function handles location logic, plus a consumer can insert a callback to be invoked before the rendering (like the default fallback with turbolinks)
If that callback returns false it will terminate rendering